src/controller.cpp

```diff
@@ -249,6 +249,40 @@ void controllerT::execute_command_internal(std::string command, size_t counter,
 	worker_counter_cv.notify_all();
 }
 
+controllerT::execute_config controllerT::generate_config(const size_t requested_cpus, const resource_selector selector) const {
+	execute_config config;
+	for (size_t m = 0; m < machine_usage.size(); ++m) {
+		const auto &mu = machine_usage[m];
+
+		switch (selector) {
+		case exclusive: {
+			// take all slots if empty
+			if (mu[0] == std::numeric_limits<size_t>::max()) {
+				for (size_t s = 0; s < SLOTS; ++s) {
+					config.emplace_back(m, s);
+				}
+			}
+			break;
+		}
+		case one_slot_per_node: {
+			// pick one slot per machine
+			for (size_t s = 0; s < SLOTS; ++s) {
+				if (mu[s] == std::numeric_limits<size_t>::max()) {
+					config.emplace_back(m, s);
+					break;
+				}
+			}
+			break;
+		}
+		}
+
+		// we found enough slots
+		if (config.size() * SLOT_SIZE >= requested_cpus) break;
+	}
+
+	return config;
```

Comment:
This could return a config containing less than the number of requested cpus without returning an error. The assert should be in this function or an error code should be returned as well.